File: src/rate-limiters/fixed-window.ts
import type { Redis } from 'ioredis';
import type { RateLimiter } from './types.js';
import type { RateLimitResult } from '../types/index.js';
import { readFileSync, existsSync } from 'node:fs';
import { resolve, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = dirname(fileURLToPath(import.meta.url));

function getLuaScript(): string {
  const p1 = resolve(__dirname, 'lua', 'fixed-window.lua');
  if (existsSync(p1)) return readFileSync(p1, 'utf-8');
  const p2 = resolve(__dirname, '..', '..', 'src', 'rate-limiters', 'lua', 'fixed-window.lua');
  if (existsSync(p2)) return readFileSync(p2, 'utf-8');
  return '';
}

const luaScript = getLuaScript();

/**
 * Fixed Window Rate Limiter
 *
 * Divides time into fixed windows and counts requests per window.
 */
export class FixedWindowLimiter implements RateLimiter {
  constructor(private redis: Redis) {}

  async check(
    apiKeyId: string,
    limit: number,
    windowSecs: number,
  ): Promise<RateLimitResult> {
    const now = Math.floor(Date.now() / 1000);
    const windowNumber = Math.floor(now / windowSecs);
    const key = `ratelimit:fixed:${apiKeyId}:${windowNumber}`;

    let result: number[];
    if (typeof (this.redis as any).fixedWindowCheck === 'function') {
      result = await (this.redis as any).fixedWindowCheck(key, limit, windowSecs) as number[];
    } else {
      result = await this.redis.eval(luaScript, 1, key, limit, windowSecs) as number[];
    }

    const [allowed, remaining, ttl, currentCount] = result;

    return {
      allowed: allowed === 1,
      remaining,
      resetAt: now + ttl,
      currentCount,
    };
  }
}

File: src/rate-limiters/lua/fixed-window.lua
-- Fixed Window Rate Limiter
-- Atomic INCR with TTL aligned to window boundaries
--
-- KEYS[1] = ratelimit:fixed:{api_key_id}:{window_number}
-- ARGV[1] = limit
-- ARGV[2] = window_secs
--
-- Returns: {allowed (0/1), remaining, ttl, current_count}

local key = KEYS[1]
local limit = tonumber(ARGV[1])
local window = tonumber(ARGV[2])

local current = redis.call('INCR', key)

if current == 1 then
  redis.call('EXPIRE', key, window)
end

-- Safety: if TTL was lost (e.g. key persisted without expiry), re-set it
local ttl = redis.call('TTL', key)
if ttl == -1 then
  redis.call('EXPIRE', key, window)
  ttl = window
end

local allowed = 0
if current <= limit then
  allowed = 1
end

local remaining = limit - current
if remaining < 0 then
  remaining = 0
end

return { allowed, remaining, ttl, current }
